### ISSUE TITLE
HttpPostRequestDecoder exception

### ISSUE THREAD

**Issue by IgorAlelekov** [netty/netty#10324]
### Actual behavior
HttpPostRequestDecoder exception

### Steps to reproduce
1. Add few lines to the Netty io.netty.example.http.helloworld.HttpHelloWorldServerHandler:
....
    @Override
    public void channelRead0(ChannelHandlerContext ctx, HttpObject msg) {
        if (msg instanceof HttpRequest) {
            HttpRequest req = (HttpRequest) msg;
// Lines to add:
if (req.method() == HttpMethod.POST) {
    HttpDataFactory factory = new DefaultHttpDataFactory(DefaultHttpDataFactory.MINSIZE);
    new HttpPostRequestDecoder(factory, req);
}
// end

2. Launch io.netty.example.http.helloworld.HttpHelloWorldServer
3. Post the following request to the 127.0.0.1:8080 address
Exception rised: io.netty.handler.codec.http.multipart.HttpPostRequestDecoder$ErrorDataDecoderException: Bad string ...

Request to post (here html tags are substituted by \&lt; and \&gt;):
&lt;s:Body xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" &gt;&lt;TransactionDetails&gt;0004486 (20%) 8333-33&lt;/TransactionDetails&gt;&lt;/s:Body&gt;

Actual request is attached

It is interestiing that exception is not rised 
A. without percent sign in the request
B. or if new line "\n" is inserted before &lt;TransactionDetails&gt; tag

### Netty version
4.1.50

### JVM version (e.g. `java -version`)
1.8.0_131

### OS version (e.g. `uname -a`)
Win-10, CentOS 7.3.1611

**Comment by IgorAlelekov** [comment]
[request.txt](https://github.com/netty/netty/files/4689911/request.txt)

**Comment by zhenjunMa** [comment]
change the code:
![image](https://user-images.githubusercontent.com/28104581/83091505-e7727800-a0cd-11ea-9868-932c723bc5c5.png)
the result:
![image](https://user-images.githubusercontent.com/28104581/83091557-ff49fc00-a0cd-11ea-8db8-fdd275294703.png)

nothing happend...

**Comment by IgorAlelekov** [comment]
Hi zhenjunMa, the result depends on the request content.
Could you please repeat the test with the **attached** request

**Comment by IgorAlelekov** [comment]
Hi, actually I added one more line. Please add it to the HttpHelloWorldServerInitializer:
....
    p.addLast(new HttpServerCodec());
// Line to add:
p.addLast(new HttpObjectAggregator(10000));
// End

**Comment by zhenjunMa** [comment]
I reproduce the exception, to summarize

case:

1. use `HttpPostRequestDecoder` after `HttpObjectAggregator`
2. request `content-type` is `application/json` or `text/plain`
3. the param contains some special charactors such as `xxx=%value`

reason:

 in HTTP `application/json` or `text/plain` only send one single string as param, not key-value , but `HttpPostStandardRequestDecoder` in `HttpPostRequestDecoder` will  call `parseBody()`  -> `parseBodyAttributes()` to decode `undecodedChunk`(holds the param),  if it found the body contains `=`,it will use string before '=' as a key  and string after '=' as a value

this has two problems

1. if `=` is the first charactor, it will throw "empty name" exception( when create `attribute`)
2. it will decode value, and think value has already `urlencoded`, so if it gets a '%', it will try 
 to get two more chars after % which must be (0~F), or it will throw 
`HttpPostRequestDecoder$ErrorDataDecoderException`

in conclusion:

is this a bug?  if so my idea to fix it is below, but how to process `undecodedChunk`?

![Screenshot from 2020-05-30 22-35-44](https://user-images.githubusercontent.com/28104581/83331034-fb80ca00-a2c5-11ea-8030-ac9a421d5a0e.png)

@normanmaurer  any idea about this?

**Comment by fredericBregier** [comment]
Maybe a dumb question, but HttpPostxxxRequestDecoder are for form decoding, meaning that passing a body with no "form" representation could not probably decode your parameters?
I didn't get if your example is to try to pass as body a JSON (so not decodable as a form, except if itself in a form attribute), or to even simpler pass a simple body with content as var=something (or whatever other form) but with no respect to a standard form value, for instance:

     Content-Disposition: form-data; name="{NOM}"[CRLF]
     [CRLF]
     {VALEUR}
